### PR TITLE
eth: Compare on chain bytecode with local copy

### LINKED ISSUE #1198: eth: Compare bytecode or swap contract with a contract on chain
When running an ETH market, the server will select a contract address which contains a swap contract. The client will receive this address when connecting to the server, and will then need to verify that the bytecode of the contract at this address matches the contract that is in the repository.

### PR DESCRIPTION
**Pull request by martonp**
Add a shell script that generates go code containing the runtime
bytecode of the ETHSwap contract. Add a function to rpcclient to
retrieve the on chain bytecode, and add a test that compares them.

Closes #1198

### CODE REVIEW

**Review by JoeGruffins — APPROVED** [review]
This is pretty perfect. Quick work of it.

Is it possible to have the github cli download the correct version of solc, run genruntime.sh and compare those files, so we would know if something changed in the hex bin? Maybe also a good idea for contract.go in that case, although that also requires a certain version of abigen I believe. Can be a different pr if others think it's a good idea.

File: dex/networks/eth/contracts/genruntime.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+ETH_SWAP_VERSION=$1
+SOLIDITY_FILE=ETHSwapV${ETH_SWAP_VERSION}.sol
+if [ ! -f ./${SOLIDITY_FILE} ]
+then
+    echo "${SOLIDITY_FILE} does not exist" >&2
+    exit 1
+fi
+
+solc --bin-runtime --optimize ${SOLIDITY_FILE} -o .
```
Comment (chappjc) [inline dex/networks/eth/contracts/genruntime.sh:20]: It's not clear to me what switches to use here.  These are the output options I see:

```
  --asm                EVM assembly of the contracts.
  --asm-json           EVM assembly of the contracts in JSON format.
  --opcodes            Opcodes of the contracts.
  --bin                Binary of the contracts in hex.
  --bin-runtime        Binary of the runtime part of the contracts in hex.
  --abi                ABI specification of the contracts.
```

The output of `--bin` is larger than `--bin-runtime`.  Which one is better to compare?

Also I'm wondering if the use of optimize (and specific compiler version) are going to make this extremely fragile.

Asking for no changes here... just a conversation.

**Review by chappjc — CHANGES_REQUESTED** [review]
Even though we're not going to be using `ethClient.getCodeAt` on the first iteration, it's a good method to have.

How do you guys control solc versions?  I have 0.8.7 and this was a pain for me.  Maybe we can allow different patch versions, at least until we are ready to lock it down?

File: client/asset/eth/rpcclient_harness_test.go [hunk 2]
```diff
@@ -842,3 +844,17 @@ func TestReplayAttack(t *testing.T) {
 		t.Fatalf("unexpected balance change of contract: want %v got %v", wantBal, bal)
 	}
 }
+
+func TestAbiEqual(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (martonp) [reply client/asset/eth/rpcclient_harness_test.go:848]: Yes, the name is not right.

File: dex/networks/eth/contracts/genruntime.sh (hunk 1 above)
Comment (martonp) [reply dex/networks/eth/contracts/genruntime.sh:20]: `--bin` returns the creation bytecode. This is what has to be sent in a transaction when you initialize a new smart contract. `--bin-runtime` returns the runtime bytecode, which is the code that ends up being stored on chain. This includes everything in the creation bytecode, except the constructor function and constructor arguments, as they will not need to be used again after initialization.

Why do you say that using a specific compiler version would make this fragile? I would think that not having a specific version would make it fragile. Optimize just applies some simplification rules to the unoptimized bytecode. This is a deterministic operation. If the same compiler version is used to optimize both what is deployed and what is used to compare, then we will get the same results.

File: dex/networks/eth/BinRuntimeV0.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
...040519150601f19603f3d011682016040523d82523d6000602084013e610552565b606091505b509091505060018115151461056657600080fd5b505050600090815260208190526040902060030155565b60006020828403121561058f57600080fd5b5035919050565b6000602082840312156105a857600080fd5b5051919050565b600080604083850312156105c257600080fd5b50508035926020909101359150565b6000806000606084860312156105e657600080fd5b833592506020840135915060408401356001600160a01b038116811461060b57600080fd5b809150509250925092565b6004811061063457634e487b7160e01b600052602160045260246000fd5b9052565b6000825160005b81811015610659576020818601810151858301520161063f565b81811115610668576000828501525b509190910192915050565b60006101008201905082518252602083015160208301526040830151604083015260608301516060830152608083015160018060a01b0380821660808501528060a08601511660a0850152505060c083015160c083015260e08301516106dc60e0840182610616565b5092915050565b8881526020810188905260408101879052606081018690526001600160a01b038581166080830152841660a082015260c08101839052610100810161072b60e0830184610616565b9998505050505050505050565b634e487b7160e01b600052602160045260246000fdfea2646970667358221220f5008532d651b31b24a7d133ff5dd0c4461e5f6339f32c9d551cb9cf81107e6b64736f6c63430008060033"
```
[the remark this answers is not in the document]
Comment (martonp) [reply dex/networks/eth/BinRuntimeV0.go:6]: That is definitely nicer, but the only reason I prefer the current technique is that it's consistent with what the abigen does (this is what generates `contract.go`). There's a really long string in there as well.

File: dex/networks/eth/contracts/genruntime.sh (hunk 1 above)
Comment (chappjc) [reply dex/networks/eth/contracts/genruntime.sh:20]: > Why do you say that using a specific compiler version would make this fragile?

Ah I was thinking about optimize and how there are different levels, but you are right, the same considerations apply there as with compiler version.  I was just referring to how identical contracts compiled with newer solc versions probably would get rejected, but I guess that's the point.

File: dex/networks/eth/BinRuntimeV0.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply dex/networks/eth/BinRuntimeV0.go:6]: That's fine.  Either we check in a ~2kb binary file to use go:embed, or we put the string right in the .go file with a shell script.
The contract bindings are actual code though, not just variables, so I see that a little differently.  It's fine this way, just a little unfortunate that it's a string when we really want bytes.

**Comment by martonp** [comment]
I haven't really thought much about managing the solidity version. I think the whole point of locking down the version was in order to be able to do these comparisons. It definitely was a pain to work with for me as well, since 0.8.6 is no longer the most recent version. Maybe a script that downloads the correct abigen and solc versions would be nice? This could be used in the github workflow that @JoeGruffins was talking about too.

**Review by chappjc — APPROVED** [review]

**Comment by JoeGruffins** [comment]
> How do you guys control solc versions? 

@chappjc I know you use arch, `sudo downgrade solidity` (after installing solidty and downgrade) and put on the ignore packages list

**Comment by chappjc** [comment]
I was thinking a less system-wide approach, but that's probably fine for my purposes. Thanks

**Comment by JoeGruffins** [comment]
KISS

But it can maybe be made better if one to were make the golang-cli script ,which would also download the correct solc to a temp location I suppose, and we could just use that to regenerate.

**Pull request merged**